### DISCUSSION TITLE
Why should I click twice on button to get data from useSWR?

### DISCUSSION THREAD

**Discussion by alamenai** [vercel/swr#2733] [Q&A]
I have a `Button` component where I get data from my backend API using [useSWR.][1]

This is my button:

```typescript

interface IButtonProps {
  address: string;
}

const AddressCheckButton = ({ address }: IButtonProps) => {
  const [shouldFetch, setShouldFetch] = useState(false);
  const dispatch = useDispatch();
  const router = useRouter();
  const url = process.env.NEXT_PUBLIC_API_URL;

  const { data: userLocation, error } = useSWR(
    shouldFetch ? url + address : null,
    (url) => fetcher(url, headers)
  );

  const handleValidate = () => {
    fetchData();
  };

  const fetchData = () => {
    try {
      setShouldFetch(true);
      if (userLocation) {
        const { zipCode, city, street, houseNumber } = userLocation;
        console.log(userLocation);

        if (zipCode && city && street && houseNumber) {
          dispatch(updateAddressId({ address }));
          dispatch(updateSolarLocation({ zipCode, city, street, houseNumber }));
          setShouldFetch(false);
          router.push(`/${address}`);
        }
      } else {
        console.log("error");
      }
    } catch (error) {
      setShouldFetch(false);
      console.log(error);
    }
  };
  return (
    <>
      <Button
        text="Code überprüfen"
        onClick={handleValidate}
        type={"ternary"}
      />
      {error?.length > 0 ? (
        <div className={styles.error}>
          <p>{error}</p>
        </div>
      ) : null}
    </>
  );
};

```
The problem is when I click ( first click ) it logs an `error' because `userLocation` is `undefined`  while the useSWR is fetching the data. The button code is synchronous and the useSWR is asynchronous, how to wait for `userLocation` to be available to run the `if` code?

  [1]: https://swr.vercel.app/docs/conditional-fetching

**Comment by harshyadavDeveloper** [comment]
You can handle this using SWR’s built-in behavior for caching and revalidation.

By default, SWR will revalidate data on focus and network reconnect. If you want to control this behavior, you can pass options like:

```js
useSWR(key, fetcher, {
  revalidateOnFocus: false,
  revalidateOnReconnect: false,
})
```

This ensures the data is only fetched when you explicitly want it.

If your issue is related to stale data, you can also use `mutate` to manually update or revalidate the cache.

**Comment by devhorizon999** [comment]
The reason you need to click twice is that `setShouldFetch(true)` does not make `userLocation` available immediately.

On the first click, this happens:

```ts
setShouldFetch(true);
if (userLocation) {
  // ...
}
